### PR TITLE
Feature 112527 refactoring map component

### PR DESCRIPTION
**Pull request by kuzkok**

### CODE REVIEW

File: addon/components/base-control.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    let leafletMap = this.get('map');
+    if (!Ember.isNone(leafletMap)) {
+      let control = this.createControl();
+      this.set('control', control);
+      leafletMap.addControl(control);
+    }
+  }),
+
+  didInsertElement() {
+    this.initControl();
```
Comment (SeregaMatin) [inline addon/components/base-control.js]: You forget to call _super(..arguments).

File: addon/components/base-layer.js [hunk 2]
```diff
@@ -99,8 +83,8 @@ export default Ember.Component.extend(
     }),
 
     /**
-      Create layer with requiredOptions and options properties, should be overriden in child classes.
-      @method
+      Create leaflet layer, should be overriden in child classes.
+      @method createLayer
      */
     createLayer() {
       assert('BaseLayer\'s `createLayer` should be overriden.');
```
Comment (SeregaMatin) [inline addon/components/base-layer.js]: Typo. Should be "overridden".

File: addon/components/flexberry-map.js [hunk 3]
```diff
@@ -77,6 +90,7 @@ export default Ember.Component.extend(
       this.set('_layer', map);
       this._addObservers();
       this._addEventListeners();
+      this.sendAction('leafletMapDidInit', map);
```
Comment (SeregaMatin) [inline addon/components/flexberry-map.js]: Component called "flexberry-map".
Don't you think that {{flexberry-map leafletMapDidInit=(action "onLeafletMapDidInit")}} looks uglier then something like that {{flexberry-map init=(action "onMapInit")}}.
I mean that action name is too long & ugly.

File: addon/components/flexberry-map.js [hunk 4]
```diff
@@ -77,6 +90,7 @@ export default Ember.Component.extend(
       this.set('_layer', map);
```
Comment (SeregaMatin) [inline addon/components/flexberry-map.js]: Why do property for leaflet-map called "_layer" why not "map"?

File: addon/components/flexberry-map.js (hunk 3 above)
Comment (kuzkok) [inline addon/components/flexberry-map.js]: It raise after creating L.Map instance. Yes, its ugly but describe exactly what happen

File: addon/components/flexberry-map.js (hunk 4 above)
Comment (kuzkok) [inline addon/components/flexberry-map.js]: For unifiying with layer's description and using in mixin leaflet-properties

File: addon/mixins/flexberry-map-load-layers.js [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+  @class FlexberryMapLoadLayersMixin
+  @uses <a href="http://emberjs.com/api/classes/Ember.Mixin.html">Ember.Mixin</a>
+ */
+export default Ember.Mixin.create({
+
+  /**
+    @return {Ember.RSVP.Promise} Promise with loading flexberry-map-layer
+    @private
+   */
+  _loadLayer(layer) {
```
Comment (SeregaMatin) [inline addon/mixins/flexberry-map-load-layers.js]: All private methods in this mixin are contextless, they don't use any methods or properties from current instance (this).
So there is no need for them to be a mixin members. They can be placed outside of the mixin.

File: addon/routes/flexberry-map.js [hunk 6]
```diff
@@ -1,62 +1,21 @@
...
-    let baseModel = null;
-    return this._super(...arguments).then(map => {
-      baseModel = map;
-      return this.loadChildLayers(map.get('rootLayer'));
-    }).then(layers => {
-      baseModel.set('layers', layers);
-      this.setIndex(layers, { index: 0 });
-      return baseModel;
-    });
+    return this.loadMapLayers(this._super(...arguments));
```
Comment (SeregaMatin) [inline addon/routes/flexberry-map.js]: Don't you think that separate calls would be clearer:
let mapPromise = this._super(...arguments);
return this.loadMapLayers(mapPromise);

File: tests/dummy/app/templates/map.hbs [hunk 7]
```diff
@@ -1,11 +1,21 @@
 <div class="ui grid">
   <div class="twelve wide column">
-    {{flexberry-map model=model zoomControl=true onMoveend=(action "updateCenter")}}
+    {{flexberry-map
+      lat=model.lat
+      lng=model.lng
+      zoom=model.zoom
+      layers=model.layers
+      zoomControl=true
+      onMoveend=(action "updateCenter")
```
Comment (SeregaMatin) [inline tests/dummy/app/templates/map.hbs]: Maybe "onUpdateCenter" ?

File: addon/mixins/flexberry-map-load-layers.js (hunk 5 above)
Comment (kuzkok) [inline addon/mixins/flexberry-map-load-layers.js]: You suggest move this methods to outside mixin definition and use its as closure functions?

File: addon/components/flexberry-map.js (hunk 3 above)
Comment (SeregaMatin) [inline addon/components/flexberry-map.js]: We don't have any other maps except leaflet inside component.
So such looong action/properties names are unnecessary.

File: addon/components/flexberry-map.js [hunk 8]
```diff
@@ -77,6 +90,7 @@ export default Ember.Component.extend(
...
       this._addObservers();
       this._addEventListeners();
+      this.sendAction('leafletMapDidInit', map);
     },
 
     didInsertElement() {
```
Comment (SeregaMatin) [inline addon/components/flexberry-map.js]: Create map's <div> inside init, and attach it to DOM later in didiInsertElement is really bad decision.
Whole component's wrapping <div> (this.$()) must be a leaflet map's container, and it should be initialized with leaflet strictly in didInsertElemet hook.

All init logic inside component itself & in nested chaild components (layers, controls) must wait (you should make a promise from map object passed to them, and force them to wait, or refactor their init logic some other way).

File: addon/mixins/flexberry-map-load-layers.js (hunk 5 above)
Comment (SeregaMatin) [inline addon/mixins/flexberry-map-load-layers.js]: Yep.

File: addon/components/flexberry-map.js (hunk 3 above)
Comment (kuzkok) [inline addon/components/flexberry-map.js]: but we have component with "map" in it name

File: addon/mixins/flexberry-map-load-layers.js (hunk 5 above)
Comment (kuzkok) [inline addon/mixins/flexberry-map-load-layers.js]: may be better remove mixin and move this functions to utils class?

Comment (kuzkok) [inline addon/mixins/flexberry-map-load-layers.js]: This methods use this.store from context.

**Pull request merged**